File: src/accessibility/SimpleOpportunity.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using DVAN.Routing;
using DVAN.Population;

namespace DVAN.Accessibility
{
    /// <summary>
    /// A simple opportunity based accessibility measure.
    /// Computes the aggregated opportunity (\sum{supply*decay}) at every population point.
    /// </summary>
    public class SimpleOpportunity
    {
        public static async Task<float[]> calcAccessibility(IPopulationView population, double[][] facilities, double[] capacities, List<double> ranges, IDistanceDecay decay, IRoutingProvider provider)
        {
            var accessibilities = new float[population.pointCount()];

            var matrix = await provider.requestTDMatrix(population, facilities, ranges, "isochrones");
            if (matrix == null) {
                return accessibilities;
            }

            for (int f = 0; f < facilities.Length; f++) {
                for (int p = 0; p < population.pointCount(); p++) {
                    float range = matrix.getRange(f, p);
                    if (range == 9999) {
                        continue;
                    }
                    float range_factor = decay.getDistanceWeight((float)range);

                    accessibilities[p] += (float)capacities[f] * (float)range_factor;
                }
            }

            float max_value = 0;
            for (int i = 0; i < accessibilities.Length; i++) {
                if (max_value == 0) {
                    max_value = accessibilities[i];
                }
                else {
                    if (max_value < accessibilities[i]) {
                        max_value = accessibilities[i];
                    }
                }
            }
            for (int key = 0; key < accessibilities.Length; key++) {
                var access = accessibilities[key];
                if (access == 0) {
                    accessibilities[key] = -9999;
                }
                else {
                    accessibilities[key] = access * 100 / max_value;
                }
            }

            return accessibilities;
        }
    }
}

File: src/main/java/org/tud/oas/accessibility/SimpleOpportunity.java
package org.tud.oas.accessibility;

import org.tud.oas.accessibility.distance_decay.IDistanceDecay;
import org.tud.oas.demand.IDemandView;
import org.tud.oas.routing.IRoutingProvider;
import org.tud.oas.routing.ITDMatrix;
import org.tud.oas.routing.RoutingOptions;
import org.tud.oas.supply.ISupplyView;

public class SimpleOpportunity {

    /**
     * Computes a simple opportunity measure. For every demand point the product of
     * reachable supply and distance decay are summed up.
     * Example formula for linear decay:
     * $A_i = \sum{S_j * (1-\frac{d_{ij}}{d_{max}})}$
     * 
     * When using the Gravity-Decay function ($d_{ij}^{-\beta}$) this method can
     * also compute the basic gravity accessibility introduced by Hansen (1959).
     * Formula:
     * $A_i = \sum{\frac{S_j}{d_{ij}^{\beta}}}$
     * 
     * @param demand   Demand locations (results contains accumulated opportunity
     *                 for every demand point).
     * @param supply   Supply locations and weights ($S_j$).
     * @param decay    Distance decay.
     * @param provider Routing API provider.
     * @param options  Mode and ranges of isochrones used in computation of
     *                 distances $d_{ij}$.
     * @return Accumulated Opportunity for every demand point.
     */
    public static float[] calcAccessibility(IDemandView demand, ISupplyView supply, IDistanceDecay decay,
            IRoutingProvider provider, RoutingOptions options) throws Exception {
        float[] accessibilities = new float[demand.pointCount()];

        ITDMatrix matrix;
        try {
            matrix = provider.requestTDMatrix(demand, supply, options);
        } catch (Exception e) {
            throw new Exception("failed to compute travel-time-matrix:" + e.getMessage());
        }

        try {
            for (int f = 0; f < supply.pointCount(); f++) {
                for (int p = 0; p < demand.pointCount(); p++) {
                    float range = matrix.getRange(f, p);
                    if (range < 0) {
                        continue;
                    }
                    float rangeFactor = decay.getDistanceWeight(range);

                    accessibilities[p] += (float) supply.getSupply(f) * rangeFactor;
                }
            }
            for (int i = 0; i < accessibilities.length; i++) {
                float access = accessibilities[i];
                if (access == 0) {
                    accessibilities[i] = -9999;
                } else {
                    accessibilities[i] = access;
                }
            }

            return accessibilities;
        } catch (Exception e) {
            e.printStackTrace();
            throw new Exception("failed to compute accessibility.");
        }
    }
}